application/comit_node/src/swap_protocols/rfc003/bob/mod.rs

```diff
@@ -1,121 +1,162 @@
+            alpha_expiry: swap_request.alpha_expiry,
+            beta_expiry: swap_request.beta_expiry,
+            secret: swap_request.secret_hash,
         };
 
-        save_state.save(start_state.clone().into());
+        let state_repo = Arc::new(RwLock::new(None));
 
         let context = Context {
             alpha_ledger_events,
             beta_ledger_events,
-            communication_events: Box::new(BobToAlice::new(Box::new(response_for_state_machine))),
-            state_repo: save_state,
+            communication_events,
+            state_repo,
         };
 
-        (
-            Box::new(Swap::start_in(start_state, context)),
-            Box::new(response_for_caller),
-        )
+        Box::new(Swap::start_in(start_state, context))
     }
 
-    pub fn create() -> (Self, Box<ResponseFuture<Self>>) {
+    pub fn new(request: Request<AL, BL, AA, BA>, secret_source: Arc<dyn SecretSource>) -> Self {
         let (sender, receiver) = oneshot::channel();
-        (
-            Bob {
-                phantom_data: PhantomData,
-                response_sender: Arc::new(Mutex::new(Some(sender))),
+        let sender = Arc::new(Mutex::new(Some(sender)));
+        let receiver = receiver.map_err(|_e| {
+            rfc003::Error::Internal(String::from(
+                "For now, it should be impossible for the sender to go 
+out of scope before the receiver",
+            ))
+        });
+        let receiver = (Box::new(receiver) as Box<ResponseFuture<AL, BL>>).shared();
+
+        State {
+            swap_communication: SwapCommunication::Proposed {
+                request,
+                response: Response { sender, receiver },
             },
-            Box::new(
-                receiver
-                    .map_err(|_e| unreachable!("For now, it should be impossible for the sender to go out of scope before the receiver") ),
-            ),
-        )
+            alpha_ledger_state: LedgerState::NotDeployed,
+            beta_ledger_state: LedgerState::NotDeployed,
+            secret_source,
+            secret: None,
+        }
     }
 
-    pub fn accept_action(&self) -> Accept<AL, BL> {
-        Accept::new(self.response_sender.clone())
+    pub fn request(&self) -> Request<AL, BL, AA, BA> {
+        match &self.swap_communication {
+            SwapCommunication::Proposed { request, .. }
+            | SwapCommunication::Rejected { request } => request.clone(),
+            SwapCommunication::Accepted { swap_accepted, .. } => swap_accepted.request.clone(),
+        }
     }
 
-    pub fn decline_action(&self) -> Decline<AL, BL> {
-        Decline::new(self.response_sender.clone())
+    pub fn response_future(&self) -> Option<Box<ResponseFuture<AL, BL>>> {
+        match &self.swap_communication {
+            SwapCommunication::Proposed { response, .. } => {
+                Some(Box::new(response.response_future()))
+            }
+            _ => None,
```

Comment:
:wrench: We could log a `warn` here with the current state because calling this in any state other than `Proposed` will likely yield an error and then this message might be a good hint to discover why.